test/new-e2e/pkg/e2e/suite_test.go

```diff
@@ -0,0 +1,157 @@
+	permanentProvisioner.On("Provision", mock.Anything, mock.Anything, mock.Anything).Return(testRawResources("myWrapper1", "permanent"), nil)
+
+	// Temp provisioner is going to be removed in some tests
+	tempProvisioner := &testProvisioner{}
+	tempProvisioner.On("ID").Return("temp")
+	tempProvisioner.On("Provision", mock.Anything, mock.Anything, mock.Anything).Return(testRawResources("myWrapper2", "temp"), nil)
+
+	Run(t, &testSuiteWithTests{permanentProvisioner: permanentProvisioner, tempProvisioner: tempProvisioner}, WithProvisioner(permanentProvisioner), WithProvisioner(tempProvisioner))
+}
+
+func (s *testSuiteWithTests) TestOrderA() {
+	s.permanentProvisioner.AssertExpectations(s.T())
+	s.permanentProvisioner.AssertNumberOfCalls(s.T(), "Provision", 1)
+	s.tempProvisioner.AssertExpectations(s.T())
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Provision", 1)
+
+	// Nothing should happen, same objects
+	s.UpdateEnv(s.permanentProvisioner, s.tempProvisioner)
+
+	s.permanentProvisioner.AssertExpectations(s.T())
+	s.permanentProvisioner.AssertNumberOfCalls(s.T(), "Provision", 1)
+	s.tempProvisioner.AssertExpectations(s.T())
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Provision", 1)
+
+	// Remove temp provisioner, destroy should be called.
+	// `Provide` will be called again on permanent provisioner
+	// The call will panic because missing resource `myWrapper2`
+	s.tempProvisioner.On("Destroy", mock.Anything, mock.Anything, mock.Anything).Return(nil)
+	s.Assert().PanicsWithError(
+		"unable to build env: *e2e.testEnv from resources for stack: e2e-testSuiteWithTests-e3e6e49fc651fcb8, err: expected resource named: Wrapper2 with key: myWrapper2 but not returned by provisioners",
+		func() { s.UpdateEnv(s.permanentProvisioner) },
+	)
+
+	s.permanentProvisioner.AssertExpectations(s.T())
+	s.permanentProvisioner.AssertNumberOfCalls(s.T(), "Provision", 2)
+	s.tempProvisioner.AssertExpectations(s.T())
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Provision", 1)
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Destroy", 1)
+
+	// As UpdateEnv failed, the `currentProvisioners` have not been updated
+	// so the next test should not call provisioners again.
+	// As we want that to happen, we'll simulate that by patching manually the `currentProvisioners`.
+	delete(s.currentProvisioners, s.tempProvisioner.ID())
+}
+
+func (s *testSuiteWithTests) TestOrderB() {
+	// In this test, the original provisioners will be called again, restoring everything
+	s.permanentProvisioner.AssertNumberOfCalls(s.T(), "Provision", 3)
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Provision", 2)
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Destroy", 1)
+}
+
+func (s *testSuiteWithTests) TestOrderC() {
+	// The provisioners were not changed, so nothing should happen
+	s.permanentProvisioner.AssertNumberOfCalls(s.T(), "Provision", 3)
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Provision", 2)
+	s.tempProvisioner.AssertNumberOfCalls(s.T(), "Destroy", 1)
+}
```

Comment:
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do look for something to sincerely praise. -->
**👏 praise**